## Author: Jamie Kostyun
## Ecological Genomics:   

### Overall Description of notebook      

This notebook will document my entries throughout the semester.


### Date started: 2018-01-24  
### Date end:   (year-month-day)    

### Philosophy   
Science should be reproducible and one of the best ways to achieve this is by logging research activities in a notebook. Because science/biology has increasingly become computational, it is easier to document computational projects in an electronic form, which can be shared online through Github.    

### Helpful features of the notebook     

**It is absolutely critical for your future self and others to follow your work.**     

* The notebook is set up with a series of internal links from the table of contents.    
* All notebooks should have a table of contents which has the "Page", date, and title (information that allows the reader to understand your work).     
* Also, one of the perks of keeping all activities in a single document is that you can **search and find elements quickly**.     
* Lastly, you can share specific entries because of the three "#" automatically creates a link when the notebook renders on github.      


<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.  


### Table of contents for 60 entries (Format is *Page: Date(with year-month-day). Title*)        
* [Page 1: 2018-01-24](#id-section1). Intro to Github, RMarkdown, and UNIX command-line
* [Page 2: 2018-01-26](#id-section2). Practice with UNIX commands
* [Page 3: 2018-01-29](#id-section3). Introduction to handling RNAseq data
* [Page 4: 2018-01-31](#id-section4). Continuing with RNAseq data: filtering and mapping to reference
* [Page 5: 2018-02-05 and 2018-02-06](#id-section5). Examining SAM file and troubleshooting 
* [Page 6: 2018-02-14](#id-section6). Using DESeq2 to assess differential gene expression
* [Page 7:](#id-section7).
* [Page 8:](#id-section8).
* [Page 9:](#id-section9).
* [Page 10:](#id-section10).
* [Page 11:](#id-section11).
* [Page 12:](#id-section12).
* [Page 13:](#id-section13).
* [Page 14:](#id-section14).
* [Page 15:](#id-section15).
* [Page 16:](#id-section16).
* [Page 17:](#id-section17).
* [Page 18:](#id-section18).
* [Page 19:](#id-section19).
* [Page 20:](#id-section20).
* [Page 21:](#id-section21).
* [Page 22:](#id-section22).
* [Page 23:](#id-section23).
* [Page 24:](#id-section24).
* [Page 25:](#id-section25).
* [Page 26:](#id-section26).
* [Page 27:](#id-section27).
* [Page 28:](#id-section28).
* [Page 29:](#id-section29).
* [Page 30:](#id-section30).
* [Page 31:](#id-section31).
* [Page 32:](#id-section32).
* [Page 33:](#id-section33).
* [Page 34:](#id-section34).
* [Page 35:](#id-section35).
* [Page 36:](#id-section36).
* [Page 37:](#id-section37).
* [Page 38:](#id-section38).
* [Page 39:](#id-section39).
* [Page 40:](#id-section40).
* [Page 41:](#id-section41).
* [Page 42:](#id-section42).
* [Page 43:](#id-section43).
* [Page 44:](#id-section44).
* [Page 45:](#id-section45).
* [Page 46:](#id-section46).
* [Page 47:](#id-section47).
* [Page 48:](#id-section48).
* [Page 49:](#id-section49).
* [Page 50:](#id-section50).
* [Page 51:](#id-section51).
* [Page 52:](#id-section52).
* [Page 53:](#id-section53).
* [Page 54:](#id-section54).
* [Page 55:](#id-section55).
* [Page 56:](#id-section56).
* [Page 57:](#id-section57).
* [Page 58:](#id-section58).
* [Page 59:](#id-section59).
* [Page 60:](#id-section60).

------
<div id='id-section1'/>
### Page 1: 2018-01-24. Notes on using Github, Rmarkdown, and the UNIX command-line.

Today, we created Github repo for the course, and started the course Notebook. 

Other goals:

* publish notebook to Github
* log into UNIX server

------
<div id='id-section2'/>
### Page 2: 2018-01-26. Practice with UNIX commands. 

Today, I practiced using UNIX commands in the Terminal, including modifying files in my home directory and connecting back to the class server. I also followed the course tutorial section about editing my settings, so as to be asked if I actually want to delete files. I think this change will be very helpful!

------
<div id='id-section3'/>
### Page 3: 2018-01-29. Introduction to handling RNAseq data

Today, we started handling RNAseq data: began learning the RNA processing pipeline, how to write a bash script, and how to interpret fastq files. 

We practiced with RNAseq data from horned beetles, where different populations have diverged in their relationship b/t male body size and horn size - i.e. different thresholds. 

First, I connected to the class server and navigated to the data directory:

      cd /data/project_data/beetles/rawdata/

...from which I examined the beginning of a fastq file:

      zcat WA_PP1_F2_AGTTCC_L003_R1_001.fastq.gz | head -n4

To examine the overall/summary quality scores for all reads in this sample (Western Australia, Female2, Paired-End1), I first created a new directory in my home directory to hold working RNAseq files. Then, I used commands from FastQ to create the quality report and have it output to my new directory:

      cd ~
      mkdir 2018_clean_reads
      fastqc /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R1_001.fastq.gz -o .

This output included a .zip file and an .html file. Because we will be using reads from both paired-end sets, I did the same thing with the other file:

      fastqc /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R2_001.fastq.gz -o .

Using cyberduck, I then moved these files to a data folder on my personal computer. Next, I opend the .html files to examine the quality scores for both sets. Overall, both sets look pretty good although PE2 has more poorer quality base calls near the end of the read. Both also have residue from the random hexamer primers at the front of the read as well as some adaptor sequence near the end - both of these issues will be fixed when we trim the reads. 

------
<div id='id-section4'/> 
### Page 4: 2018-01-31. Continuing with RNAseq data: filtering and mapping to reference

Today, we continued dealing with beetle RNAseq data. After examining the overall quality with the FastQ reports, we used the program Trimmomatic to trim and filter reads. To do so, we first copied an example script for Trimmomatic from the course materials to our home directory (into myscripts directory):

      cp /data/scripts/trim_example.sh ~/myscripts/
      
We then used the program Vim to edit the example script within the command window:

      vim trim_example.sh
      
Inside the vim edit window, this is what the example script looked like:

      #!/bin/bash
        java -classpath /data/popgen/Trimmomatic-0.33/trimmomatic-0.33.jar org.usadellab.trimmomatic.TrimmomaticPE \
                -threads 1 \ 
                -phred33 \
                 /data/project_data/beetles/rawdata/WA_PP1_YOURSAMPLE_R1.fastq.gz \
                 /data/project_data/beetles/rawdata/WA_PP1_YOURSAMPLE_R2.fastq.gz \
                 ~/cleanreads/"YOURSAMPLE_R1_clean_paired.fa" \
                 ~/cleanreads/"YOURSAMPLE_R1_clean_unpaired.fa" \
                 ~/cleanreads/"YOURSAMPLE_R2_clean_paired.fa" \
                 ~/cleanreads/"YOURSAMPLE_R2_clean_unpaired.fa" \
                 ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 \
                 LEADING:28 \
                TRAILING:28 \
                SLIDINGWINDOW:6:28 \
                HEADCROP:12 \
                MINLEN:35 \
                
This is specifying that it's a Java based program, where to find it (in the shared class space in data/popgen), and that we'll be using it to analyze paired end data. After the -phred33 \, we specify the input files - there are two because we're doing paired end analysis. Next, it lists the output files for the cleaned/trimmed reads. Finally, it lists the steps of functions to perform on the raw reads. Here's more info on these steps:

ILLUMINACLIP: Cut adapter and other illumina-specific sequences from the read.
SLIDINGWINDOW: Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold.
LEADING: Cut bases off the start of a read, if below a threshold quality
TRAILING: Cut bases off the end of a read, if below a threshold quality
CROP: Cut the read to a specified length
HEADCROP: Cut the specified number of bases from the start of the read
MINLEN: Drop the read if it is below a specified length

To edit the script in vim, need to type "i" for INSERT. After changes are complete, hit escape key to exit edit mode. To enter general commands, type ":" followed by the command. For example, ":w" means write the changes to the file, and ":q" means quit vim.

After editing, this is what my Trimmomatic script looked like:

      #!/bin/bash
        java -classpath /data/popgen/Trimmomatic-0.33/trimmomatic-0.33.jar org.usadellab.trimmomatic.TrimmomaticPE \
                      -threads 1 \
                      -phred33 \
                       /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R1_001.fastq.gz \
                       /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R2_001.fastq.gz \
                       ~/2018_clean_reads/"WA_PP1_F2_R1_clean_paired.fa" \
                       ~/2018_clean_reads/"WA_PP1_F2_R1_clean_unpaired.fa" \
                       ~/2018_clean_reads/"WA_PP1_F2_R2_clean_paired.fa" \
                       ~/2018_clean_reads/"WA_PP1_F2_R2_clean_unpaired.fa" \
                       ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 \
                       LEADING:28 \
                   TRAILING:28 \
                   SLIDINGWINDOW:6:28 \
                   HEADCROP:12 \
                   MINLEN:35 \

I then changed the filename, and also changed the user permissions so I could execute it. Here, r means readable, w means writeable, and x means executable:

      mv trim_example.sh trim_Fl.sh
      chmod u+x trim_Fl.sh
      
While still in the myscripts directory, I then executed the script:

      ./trim_Fl.sh 
      
Output inside terminal window:

      TrimmomaticPE: Started with arguments: -threads 1 -phred33        /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R1_001.fastq.gz /data/project_data/beetles/rawdata/WA_PP1_F2_AGTTCC_L003_R2_001.fastq.gz /users/j/k/jkostyun/2018_clean_reads/WA_PP1_F2_R1_clean_paired.fa /users/j/k/jkostyun/2018_clean_reads/WA_PP1_F2_R1_clean_unpaired.fa /users/j/k/jkostyun/2018_clean_reads/WA_PP1_F2_R2_clean_paired.fa /users/j/k/jkostyun/2018_clean_reads/WA_PP1_F2_R2_clean_unpaired.fa ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 LEADING:28 TRAILING:28 SLIDINGWINDOW:6:28 HEADCROP:12 MINLEN:35
      Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'
      ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
      Input Read Pairs: 16000000 Both Surviving: 10714663 (66.97%) Forward Only Surviving: 2565021 (16.03%) Reverse       Only Surviving: 734046 (4.59%) Dropped: 1986270 (12.41%)
      TrimmomaticPE: Completed successfully
      
To make sure that the script actually did what we thought it was going to, we ran FastQ on the clean reads file:

      cd 2018_clean_reads
      fastqc ~/2018_clean_reads/WA_PP1_F2_R1_clean_paired.fa
      fastqc ~/2018_clean_reads/WA_PP1_F2_R2_clean_paired.fa
      
Once satisfied with the improved quality, we then started preparing to map the clean reads against the reference. The reference for O. taurus is already available in the shared course space. To first index the file (only needs to be done once), we used the program bwa:

      bwa index /data/project_data/beetles/reference/OTAU.fna
      
Then we mapped clean reads to the reference, and output (sam file).

      bwa mem /data/project_data/beetles/reference/OTAU.fna  ~/2018_clean_reads/WA_PP1_F2_R1_clean_paired.fa ~/2018_clean_reads/WA_PP1_F2_R2_clean_paired.fa > WA_PP1_F2_bwamem.sam

------
<div id='id-section5'/>
### Page 5: 2018-02-05 and 2018-02-06. Examining SAM file and troubleshooting

Today at the beginning of class, we used the -tail function to start examining our SAM file:
      
      tail -n 100 WA_PP1_F2_bwamem.sam > tail.PP1_F2.sam
      vim tail.PP1_F2.sam
      
However, when I went to examine it, it was an empty file. So was my SAM file. Clearly there was a problem with my mapping command/output, so I followed along in class using a SAM file available in the shared class space.

But now I realized that my actual SAM file was just in another folder, and the empty one was from a first attempt at mapping where I got an error message. So I moved the correct file to the correct folder. Problem solved.

After looking over the vim field, just to get a sense of what's included in a SAM file, we assessed overall mapping quality:

      samtools flagstat WA_PP1_F2_bwamem.sam
      
Output inside terminal window:

      21506273 + 0 in total (QC-passed reads + QC-failed reads)
      0 + 0 secondary
      76947 + 0 supplementary
      0 + 0 duplicates
      17749530 + 0 mapped (82.53% : N/A)
      21429326 + 0 paired in sequencing
      10714663 + 0 read1
      10714663 + 0 read2
      17100460 + 0 properly paired (79.80% : N/A)
      17369268 + 0 with itself and mate mapped
      303315 + 0 singletons (1.42% : N/A)
      183336 + 0 with mate mapped to a different chr
      151063 + 0 with mate mapped to a different chr (mapQ>=5)
      
Next we used a python script to count reads mapping to different contigs/"genes":

      python countxpression_PE.py 20 35 ~/mydata/sam/PP1_F2_countstatssummary.txt ~/mydata/sam/WA_PP1_F2_bwamem.sam
      
However, the file only contains 0 for all reads, suggesting there is something wrong with the script. Indeed, it seems that the mem function in BWA (which we used for mapping) may have somewhat different flags than the align function (which is what the script was written for).
      
------
<div id='id-section6'/>
### Page 6: 2018-02-07. Exploring expression differences using DESeq2.

Today, we continued our exploration and analysis of beetle RNAseq data. Last time, the script we were going to use to get count data didn't work on the BWA mem function. So Melissa provided an alternate file for us that used the BWA align function to get counts, and then a bash script to combine read counts from all individuals into a single file.

First, we copied the relevant files from the course space. Typically I would just use CyyberDuck, but here I tried this:

    scp jkostyun@pbio381.uvm.edu:/data/project_data/beetles/counts/* .
    
Fo this command to work, I was in the directory on my computer that I wanted to transfer the files to (indicated by the "."). This copied two files "allcountsdataRN_noIT.txt" and "cols_data_noIT.txt".

Next, we loaded the DESeq2 package in R(Studio) and started writing a script to use for differential expression analyses. 

      #set working directory
      library("DESeq2")
      library("ggplot2")

      countsTable = read.delim('allcountsdataRN_noIT.txt', header=T, stringsAsFactors = T, row.names=1)
      countData = as.matrix(countsTable)
      head(countsData)

      conds = read.delim('cols_data_noIT.txt', header=T, stringsAsFactors = T, row.names=1)
      head(conds)
      colData = as.data.frame(conds)

      ################

      dds = DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~ devstage + sex + population)
      # DEseq2 interprets models as testing differences b/t last predictor accounting for previous predictors

      dim(dds) #gives summary of input: 17483    48

      dds1 = dds[rowSums(counts(dds)) > 1]  #revises dds to exclude transcripts with 0 reads, as previous analysis        included all transcripts
      dim(dds1) #gives updated summary: 16851    48

      dds1 = DESeq(dds1, modelMatrixType = "standard") #this is the primary command to compare transcript counts

      resultsNames(dds1)
      #[1] "Intercept" "devstage_L3L_vs_AD4" "devstage_PD1_vs_AD4" "devstage_PP1_vs_AD4" "sex_M_vs_F"         
      #[6] "population_WA_vs_NC"

      res = results(dds1)  #this aggregates output of DESeq command
      res = res[order(res$padj),] #this re-orders output by adjusted p-value
      head(res)

      #log2 fold change (MAP): population WA vs NC 
      #Wald test p-value: population WA vs NC 
      #DataFrame with 6 rows and 6 columns
      # baseMean log2FoldChange     lfcSE      stat       pvalue         padj
      # <numeric>      <numeric> <numeric> <numeric>    <numeric>    <numeric>
      # OTAU008667-RA 231.87111     -0.7803753 0.1132661 -6.889755 5.588875e-12 4.220159e-08
      # OTAU012562-RA 251.77436     -0.7828048 0.1135355 -6.894800 5.394059e-12 4.220159e-08
      # OTAU011160-RA  10.24152     -1.0863185 0.1836927 -5.913781 3.343429e-09 1.470408e-05
      # OTAU012716-RA 188.87768      1.0137238 0.1721500  5.888609 3.894604e-09 1.470408e-05
      # OTAU002976-RA 998.42315     -0.6159655 0.1075972 -5.724735 1.035950e-08 3.128984e-05
      # OTAU014686-RA 603.66091     -0.8168345 0.1468336 -5.562995 2.651835e-08 6.674668e-05

      summary(res)
      # out of 16851 with nonzero total read count
      # adjusted p-value < 0.1
      # LFC > 0 (up)     : 264, 1.6% 
      # LFC < 0 (down)   : 245, 1.5% 
      # outliers [1]     : 133, 0.79% 
      # low counts [2]   : 1616, 9.6% 
      # (mean count < 1)
      # [1] see 'cooksCutoff' argument of ?results
      # [2] see 'independentFiltering' argument of ?results



------
<div id='id-section7'/>
### Page 7: 2018-02-14. Using DESeq2 to assess differential gene expression. 

Today, we continued using DESeq2 to examine differences in gene expression between beetle populations, sex, and developmental stage. We also started examining different tools for visualizing these data.

Here is the additional code:

      ###Data Visualization###

      plotMA(res, main="DESeq2", ylim=c(-2,2)) #this plots population comparison
      abline(h=c(-1,1), col="blue", lwd = 2)

      # sex effect?

      res_sex <- results(dds1, name="sex_M_vs_F", alpha=0.05) #need to rerun model above looking at sex instead of population
      plotMA(res_sex, main="DESeq2", ylim=c(-15,15))
      abline(h=c(-1,1), col="blue", lwd = 2)

      #PCA to get a global view of gene expression
      rld <- rlog(dds1, blind=FALSE) #takes a while
      data <- plotPCA(rld, intgroup=c("population","devstage","sex"), returnData=TRUE)
      percentVar <- round(100 * attr(data, "percentVar"))

      data$devstage <- factor(data$devstage, levels=c("L3L","PP1","PD1","AD4"), labels=c("L3L","PP1","PD1","AD4"))

      ggplot(data, aes(PC1, PC2, color=sex, shape=devstage)) +
        geom_point(size=4, alpha=0.85) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance")) +
        theme_minimal()

      ggplot(data, aes(PC1, PC2, color=population, shape=devstage)) +
        geom_point(size=4, alpha=0.85) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance")) +
        theme_minimal()


      #this section will make a heatmap of all the samples vs all samples

      library("RColorBrewer")
      library("pheatmap")

      sampleDists = dist(t(assay(rld)))
      sampleDistMatrix = as.matrix(sampleDists)
      rownames(sampleDistMatrix) = paste(rld$population, rld$devstage, rld$sex, sep="-")
      colnames(sampleDistMatrix) = NULL
      colors = colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

      pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists, clustering_distance_cols = sampleDists, col=colors)

      #examining specific transcripts more closely

      d = plotCounts(dds1, gene="OTAU014686-RA", intgroup= (c("population", "sex", "devstage")), returnData = T)
      p = ggplot(d, aes(x=devstage, y=count, shape=sex, color=population)) + theme_minimal() +   theme(text=element_text(size=20), panel.grid.major=element_line(color="grey"))
      p = p + geom_point(position=position_jitter(w=0.3, h=0), size=3) + scale_x_discrete(limits=c("L3L","PP1","PD1","AD4"))


      ###Save outputs###

      write.csv(res, file="DGE_NCvsWA_pop.csv", row.names=T, quote=F) #this outputs results as csv file

      neglogpval = as.matrix(-log(res$pvalue))   #script for GO Mann-Whitney test requires this format
      head(neglogpval)
      res_negpval = cbind(row.names(res), neglogpval)
      colnames(res_negpval)=c("gene", "neglogpval")

      write.csv(res_negpval, file="DGE_NCvsWA_pop_neglogpval.csv.csv", row.names=F, quote=F)


Next, we started the process for GO term assignment and assessment of significant category enrichment. For this, the required files (scripts and databases) were provided to us in the shared course space. To use them, I moved them onto my computer to run in R. I also moved the DESeq2 output file into the same directory on my computer, making sure it was in the right format. 

      input="DGE_NCvsWA_pop_neglogpval.csv" # two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either sgnificant or not).
      ## Make sure this is saved as Unix (LF) - open in TextWrangler, save as, change from Classic Mac to Unix (LF)!!!
      goAnnotations="gene_annotation_only.tab" # two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
      goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
      goDivision="BP" # either MF, or BP, or CC
      source("gomwu.functions.R")

      # Calculating stats. It takes ~3 min for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previos runs first.
      gomwuStats(input, goDatabase, goAnnotations, goDivision, 
      perlPath="perl", # replace with full path to perl executable if it is not in your system's PATH already
      largest=0.1,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
      smallest=5,   # a GO category should contain at least this many genes to be considered
      clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. 
      Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead
      )


------
<div id='id-section8'/>
### Page 8:
------
<div id='id-section9'/>
### Page 9:
------
<div id='id-section10'/>
### Page 10:
------
<div id='id-section11'/>
### Page 11:
------
<div id='id-section12'/>
### Page 12:
------
<div id='id-section13'/>
### Page 13:
------
<div id='id-section14'/>
### Page 14:
------
<div id='id-section15'/>
### Page 15:
------
<div id='id-section16'/>
### Page 16:
------
<div id='id-section17'/>
### Page 17:
------
<div id='id-section18'/>
### Page 18:
------
<div id='id-section19'/>
### Page 19:
------
<div id='id-section20'/>
### Page 20:
------
<div id='id-section21'/>
### Page 21:
------
<div id='id-section22'/>
### Page 22:
------
<div id='id-section23'/>
### Page 23:
------
<div id='id-section24'/>
### Page 24:
------
<div id='id-section25'/>
### Page 25:
------
<div id='id-section26'/>
### Page 26:
------
<div id='id-section27'/>
### Page 27:
------
<div id='id-section28'/>
### Page 28:
------
<div id='id-section29'/>
### Page 29:
------
<div id='id-section30'/>
### Page 30:
------
<div id='id-section31'/>
### Page 31:
------
<div id='id-section32'/>
### Page 32:
------
<div id='id-section33'/>
### Page 33:
------
<div id='id-section34'/>
### Page 34:
------
<div id='id-section35'/>
### Page 35:
------
<div id='id-section36'/>
### Page 36:
------
<div id='id-section37'/>
### Page 37:
------
<div id='id-section38'/>
### Page 38:
------
<div id='id-section39'/>
### Page 39:
------
<div id='id-section40'/>
### Page 40:
------
<div id='id-section41'/>
### Page 41:
------
<div id='id-section42'/>
### Page 42:
------
<div id='id-section43'/>
### Page 43:
------
<div id='id-section44'/>
### Page 44:
------
<div id='id-section45'/>
### Page 45:
------
<div id='id-section46'/>
### Page 46:
------
<div id='id-section47'/>
### Page 47:
------
<div id='id-section48'/>
### Page 48:
------
<div id='id-section49'/>
### Page 49:
------
<div id='id-section50'/>
### Page 50:
------
<div id='id-section51'/>
### Page 51:
------
<div id='id-section52'/>
### Page 52:
------
<div id='id-section53'/>
### Page 53:
------
<div id='id-section54'/>
### Page 54:
------
<div id='id-section55'/>
### Page 55:
------
<div id='id-section56'/>
### Page 56:
------
<div id='id-section57'/>
### Page 57:
------
<div id='id-section58'/>
### Page 58:
------
<div id='id-section59'/>
### Page 59:
------
<div id='id-section60'/>
### Page 60:

------